zipline/data/data_portal.py

```diff
@@ -0,0 +1,474 @@
+        if len(adjustments_list) == 0:
+            return
+
+        # advance idx to the correct spot in the adjustments list, based on
+        # when the window starts
+        idx = 0
+
+        while idx < len(adjustments_list) and dts_in_window[0] >\
+                adjustments_list[idx][0]:
+            idx += 1
+
+        # if we've advanced through all the adjustments, then there's nothing
+        # to do.
+        if idx == len(adjustments_list):
+            return
+
+        while idx < len(adjustments_list):
+            adjustment_to_apply = adjustments_list[idx]
+
+            if adjustment_to_apply[0] > dts_in_window[-1]:
+                break
+
+            range_end = dts_in_window.searchsorted(adjustment_to_apply[0])
+            if multiply:
+                window_data[0:range_end] *= adjustment_to_apply[1]
+            else:
+                window_data[0:range_end] /= adjustment_to_apply[1]
+
+            idx += 1
+
+    def _get_adjustment_list(self, asset, adjustments_dict, table_name):
+        """
+        Internal method that returns a list of adjustments for the given sid.
+
+        Parameters
+        ----------
+        asset : Asset
+            The asset for which to return adjustments.
+
+        adjustments_dict: dict
+            A dictionary of sid -> list that is used as a cache.
+
+        table_name: string
+            The table that contains this data in the adjustments db.
+
+        Returns
+        -------
+        adjustments: list
+            A list of [multiplier, pd.Timestamp], earliest first
+
+        """
+        if self._adjustment_reader is None:
+            return []
+
+        sid = int(asset)
+
+        try:
+            adjustments = adjustments_dict[sid]
+        except KeyError:
+            adjustments = adjustments_dict[sid] = self._adjustment_reader.\
+                get_adjustments_for_sid(table_name, sid)
+
+        return adjustments
+
+    def _check_is_currently_alive(self, asset, dt):
+        if dt is None:
+            dt = self.current_day
+
+        sid = int(asset)
+
+        if sid not in self._asset_start_dates:
+            self._get_asset_start_date(asset)
+
+        start_date = self._asset_start_dates[sid]
+        if self._asset_start_dates[sid] > dt:
+            raise NoTradeDataAvailableTooEarly(
+                sid=sid,
+                dt=dt,
+                start_dt=start_date
+            )
+
+        end_date = self._asset_end_dates[sid]
+        if self._asset_end_dates[sid] < dt:
+            raise NoTradeDataAvailableTooLate(
+                sid=sid,
+                dt=dt,
+                end_dt=end_date
+            )
+
+    def _get_asset_start_date(self, asset):
+        self._ensure_asset_dates(asset)
+        return self._asset_start_dates[asset]
+
+    def _get_asset_end_date(self, asset):
+        self._ensure_asset_dates(asset)
+        return self._asset_end_dates[asset]
+
+    def _ensure_asset_dates(self, asset):
```

Comment:
The `AssetFinder` class has built-in functionality for efficiently getting the start dates of every asset we know about.  I'd suggest grabbing that information all at once rather than having to constantly chain through these methods.